### ISSUE TITLE
Do not allow random values in opts dictionaries

### ISSUE THREAD

**Issue by ZyX-I** [neovim/neovim#8284]
In the current state there are various functions which take a dictionary, use some values from it and ignore the rest. In some cases they even use that very same dictionary as `self` for callbacks found inside. I propose to stop this practice and error out if dictionary happens to contain unknown values. For the dictionaries used for `self` I propose to have some prefix for user data (e.g. `u_`) which is the only one allowed to be ignored. Reasoning:

1. Having errors for unknown keys allows testing presence of certain features in EAFP style without fearing of unintended side-effects in case function with `opts` changes state.
2. Errors present uniform way of testing for features, without errors it would be needed to construct a test based on what function does which may be convoluted.
3. If programmer bothered to supply some value to a function then he did it for a reason. If key is simply ignored function may deviate from intended behaviour and produce bugs which are hard to find.

To implement such behaviour it is logical to create a function or macros similar to Python’s [PyArg_ParseTupleAndKeywords](https://docs.python.org/3/c-api/arg.html#c.PyArg_ParseTupleAndKeywords) which may also be made responsible for parsing positional parameters, otherwise adding code implementing such behaviour will result in too much of a hassle. (Actually two functions: one for Dictionary, one for typval_T.)

Same reasoning applies to string flags, based on what I know only `writefile()` does not ignore unknown flags (and only because I modified it to not do so).

Proposed change, of course, destroys backwards compatibility. But for functions other then ones taking callbacks this should be harmless, and functions which do were introduced in Neovim and there is no Vim compatibility in any case. But this is a thing which needs to be decided before the 1.0 release as there will be backwards compatibility to keep at that time.

**Comment by ZyX-I** [comment]
The “mostly work” is a problem. Imagine that you are using `writefile()` in a plugin and want to append to a file. You use `a` flag with an old version of Vim and function succeeds, except for one thing: it does not append, it overwrites. Best case of that “mostly working” is that you lost all log entries, except one which made some problem not debuggable. Worst case is that you lost all data entries, except one and plugin ceased to function properly as it needs to read the data back and data is not there.

**Comment by justinmk** [comment]
I find partials really awkward but either way is fine with me. I think `userdata` is a useful convention in general though.

**Comment by ZyX-I** [comment]
`userdata` is only needed for callbacks to store something into. If to have `self` partials are required there would not be any need to do so as they come with a different dictionary.

**Comment by bfredl** [comment]
To be able to build a prototype dict with options and handlers, and then send in a copy to `jobstart()` etc is handy. Therefore I think something like `userdata` should still be allowed.

**Comment by ZyX-I** [comment]
@bfredl Prototyping just as well may be done in another dict, I do not see a reason to do so exactly in options dictionary. Also using a separate dictionary rather then accessing anything via `self.userdata` is easier to write. And note that options dictionary in this case is essentially read-only: it is read once on invocation and changes are ignored after that. Userdata breaks this, without that I can write in documentation that user should not expect any changes to options to have any effect. Why should one expect that modifying `userdata` has effect on next callback calls and modifying `on_stdout` has no effect at all, except for maybe confusing other dictionary users?

**Comment by ZyX-I** [comment]
> Well, for the simple reason you can specify options in the prototype. I'm not sure why this is pattern that should be forbidden _a priori_.

If something should not be done in actual code then no need to allow that in prototype as well. Or it will be done outside of prototyping for sure.

> Not sure of that, you have to first bind all you callbacks to your private self dict before placing them in the options dict, which seems a bit clumsy, and the `copy()` pattern doesn't work anymore. Or if you think it does, please show how.

For copy you just copy that private dictionary. Basically

```VimL
let opts = {}
function opts.on_stdout(...)
endfunction
…
let ocpy = copy(opts)
call job_start(['echo', 'abc'], ocpy)
```

vs

```VimL
let slf = {}
function slf.on_stdout(...)
endfunction
…
let slfcpy = copy(slf)
call job_start(['echo', 'abc'], {'on_stdout': slfcpy.on_stdout})
```

I do not really see a big deal here, it maybe is even better as you see used options in `job_start()` invocation rather then needing to scroll somewhere else. If there is really a need to have a copy of options too and unable to have dictionary literal for some reason then you would certainly have to resolve to holding lambda accepting `slfcpy`, rather then options dictionary itself though.

> I have not ever seen this being a confusion all this time we have allowed this. (and I have seen many confusions regarding job control). But if this truly is a concern, `userdata` can of course be read-only; if it is the _only_ user key it will likely be a internally mutable dict anyway.

Which would still stand out, not to mention the fact that making `self` dictionary *actually* read-only requires VAR_FIXED and may tamper other uses. Though what I actually wanted to say is more “oneshot” then “read-only”: value is examined and applied only once, after that you are free to do with the dictionary and values whatever you like to.

> Also, currently streams can be added to the options/`self` dict for `on_exit` (or `jobwait()`), in buffered mode with no dedicated stream callbacks. If we really want to hammer in that options dict is read-only, we might want to change this, to specify the other dict or something. (it is not in a release yet, so it wouldn't at least be release-breaking to change it)

Did not really understand that paragraph. How do you add streams, like the above examples with `copy()` removed? I see that would work in both examples identically.

**Comment by bfredl** [comment]
>  it maybe is even better as you see used options in job_start() invocation rather then needing to scroll somewhere else

Fair enough, that might be worth the indirection.

> Which would still stand out,

Well, partials also contain inner mutability for sure. 

> How do you add streams, like the above examples with copy() removed?

By setting the `stdout_buffered` option but not `on_stdout`. Then it is available as `self.stdout` to `on_exit`, which allows to use only one callback and not three if one only is interested in final result. This will no longer be true with indirect self, unless we change it. (it can also be used with `jobwait()` to emulate a more flexible `systemlist()`, though this is not affected by this change)

**Comment by ZyX-I** [comment]
`self.stdout` is a thing which looks inconsisent with the rest of the API: none of other variants *require* using dictionary function in a callback. And there is a number of less inconsistent alternatives:

- Pass collected data as another argument (dictionary since there are stdout and stderr).
- Transform event `data` argument (currently holds exit code) into a dictionary and pass everything there. Probably a good idea anyway since it allows passing other additional data: e.g. could have signal name here if job was finished by a signal (`rc` shell authors actually found it useful to transform some exit codes into signal names, but I would not like exit codes be either numbers or strings).
- Add `jobrecv()` function which will flush collected data. Using that is valid up until exit callback has finished.

Or really do not actually bother, I see that documented behaviour is invoking `on_stdout` with data and not use `stdout` key. I would more like to see `{'on_stdout': Callback, 'stdout_buffered': 1, 'on_stderr': 'stdout'}` working: this is *not* invoking one callback for stdout and stderr, this is making stdout and stderr use one and the same pipe (BTW, Vim does allow such binding with `'err_io': 'out'`; I find that useful despite that using separate key looks unneccessary here). Should be sufficient in many cases, though otherwise it would be three callbacks where two store data and one is `on_exit`.

**Comment by bfredl** [comment]
> none of other variants require using dictionary function in a callback.

Well it doesn't require it  (you just need any reference to the dict), it is just the most convenient way. It can also work with `jobwait()` and no callback at all. For this reason `chanrecv()` sounds best of the alternatives, but also allow it to use after `jobwait()`. Though a downside is another root for leaked value: If `stdout_buffered` is used with `jobwait()` and no callbacks, but the value is never `chanrecv()`, the buffer will be leaked. But I guess we can free leaked buffer when the main loop is reached after `jobwait()` (so it is guaranteed to work in `on_exit` or just after `jobwait()` but not later)

> 'on_stderr': 'stdout'

Would also be nice to have.

**Comment by lewis6991** [comment]
Improving vimscript is no longer a goal.

**Comment by justinmk** [comment]
@lewis6991 this topic applies generally to the API and Lua stdlib too. But it's unlikely that we would implement this because it would make backwards compatible options awkward to use.
